FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmLoginViewModel.kt

```diff
@@ -0,0 +1,160 @@
+  val ekmLiveData: MutableLiveData<Result<ApiResponse>?> = MutableLiveData()
+
+  fun fetchPrvKeys(account: String, uuid: String, tokenId: String) {
+    ekmLiveData.value = Result.loading()
+    val context: Context = getApplication()
+
+    viewModelScope.launch {
+      ekmLiveData.value = Result.loading(progressMsg = context.getString(R.string.loading))
+      try {
+        val loginResult = repository.login(
+          context,
+          LoginRequest(ApiName.POST_LOGIN, LoginModel(account, uuid), tokenId)
+        )
+
+        when (loginResult.status) {
+          Result.Status.ERROR, Result.Status.EXCEPTION -> {
+            ekmLiveData.value = loginResult
+            return@launch
+          }
+
+          Result.Status.SUCCESS -> {
+            if (loginResult.data?.isVerified == true) {
+              ekmLiveData.value = Result.loading(
+                progressMsg = context.getString(R.string.loading_domain_rules)
+              )
+              val domainRulesResult = repository.getDomainRules(
+                context,
+                DomainRulesRequest(ApiName.POST_GET_DOMAIN_RULES, LoginModel(account, uuid))
+              )
+
+              if (domainRulesResult.status == Result.Status.SUCCESS) {
+                if (domainRulesResult.data?.orgRules == null) {
+                  ekmLiveData.value = Result.exception(
+                    IllegalArgumentException("${OrgRules::class.java.simpleName} is not specified")
+                  )
+                  return@launch
+                }
+
+                if (!domainRulesResult.data.orgRules.usesKeyManager() ||
+                  !domainRulesResult.data.orgRules.mustAutoImportOrAutoGenPrvWithKeyManager()
+                ) {
+                  ekmLiveData.value = Result.exception(
+                    EkmNotSupportedException(domainRulesResult.data.orgRules)
+                  )
+                  return@launch
+                }
+
+                val notSupportedCombination =
+                  checkNotSupportedOrgRulesCombination(domainRulesResult.data.orgRules)
+
+                if (notSupportedCombination.isNotEmpty()) {
+                  ekmLiveData.value = Result.exception(
+                    OrgRulesCombinationNotSupportedException(
+                      orgRules = domainRulesResult.data.orgRules,
+                      combination = notSupportedCombination
+                    )
+                  )
+                  return@launch
+                }
+
+                val ekmPrivateResult = repository.getPrivateKeysViaEkm(
+                  context = context,
+                  ekmUrl = domainRulesResult.data.orgRules.keyManagerUrl
+                    ?: throw java.lang.IllegalArgumentException("key_manager_url is empty"),
+                  tokenId = tokenId
+                )
+
+                ekmLiveData.value = ekmPrivateResult
+              } else {
+                ekmLiveData.value = domainRulesResult
+              }
```

Comment:
This is definitely some confusing code (to my eyes - maybe not used to it). That the same LiveData object is reused for very different types of data or results. I feel like with async/await and suspend functions this would have been very clear and straight forward, but I'm not sure if UI code is somehow not compatible with that?

Reply:
Or could you have more than one livedata? That would be clearer, like `orgRulesLiveData` and `ekmLiveData` ?

Reply:
Or could the liveData have more than one value? Like `liveData.ekmResult` and `liveData.orgRulesResult`?

Reply:
> This is definitely some confusing code (to my eyes - maybe not used to it). That the same LiveData object is reused for very different types of data or results. I feel like with async/await and suspend functions this would have been very clear and straight forward, but I'm not sure if UI code is somehow not compatible with that?

`Draft` state

Reply:
> Or could you have more than one livedata? That would be clearer, like `orgRulesLiveData` and `ekmLiveData` ?

Of course, we can have that. it makes sense if we are going to fetch OrgRules somewhere else. I mean splitting these things. 

What task do I have? - with "sign-in" option I should login, make some checking and receive private keys. All of that looks like a simple operation: `do request -> receive response`. we can separate all steps like have the following

1. `LoginViewModel`
2. `OrgRulesViewModel`
3. `EkmViewModel`

But in that case, we will have some additional code(currently I need to have only one subscriber in the code, for one LiveData) and places where we can have synchronization issues. For me, it makes sense if we are going to use that ViewModel(s) in some other places.

Reply:
Yes, it definitely makes sense to split it, because:

1.    LoginViewModel - this always has to happen
2.    OrgRulesViewModel - this will soon grab OrgRules either from fes.example.com or from flowcrypt.com/api. So there will be additional logic to try FES first, then fallback to domain, or skip all of OrgRules completely based on the domain. See https://github.com/FlowCrypt/flowcrypt-android/issues/1202 so I think it deserves it's own encapsulation.
3.    EkmViewModel - this happens only if OrgRulesViewModel decides so, else it will go to some other standard setup view models.